File: lib/dice.js
// Deterministic hourly dice — same window for all clients simultaneously.
// Current product meaning:
//   r1 → activation offset within the hour (1–2699 s, ends by 59:59)
//   r3 → commission modifier: continuous random in [−0.50, +0.50] at 1% precision
// The UI uses a static 🎲 glyph plus the percentage modifier.
// A face lane still exists internally for backward compatibility, but it has no gameplay meaning.

export const DICE_FACES  = ['⚀', '⚁', '⚂', '⚃', '⚄', '⚅']

// negative modifier → cheaper commissions (cyan), positive → pricier (orange)
const COLOR_CHEAP  = '#22d3ee'
const COLOR_PRICEY = '#fb923c'

function seededRand(n) {
  let s = (n ^ 0xdeadbeef) >>> 0
  s = Math.imul(s ^ (s >>> 16), 0x45d9f3b) >>> 0
  s = Math.imul(s ^ (s >>> 16), 0x45d9f3b) >>> 0
  return ((s ^ (s >>> 16)) >>> 0) / 0x100000000
}

function modifierColor(modifier) {
  return modifier < 0 ? COLOR_CHEAP : COLOR_PRICEY
}

export function getDiceWindowForHour(hourStartMs) {
  const seed = Math.floor(hourStartMs / 3_600_000)
  const r1 = seededRand(seed * 1664525  + 1013904223)  // activation offset
  const r2 = seededRand(seed * 22695477 + 1013904223)  // legacy face lane (unused in product semantics)
  const r3 = seededRand(seed * 6364136  + 1442695041)  // modifier value

  const startSecond = Math.floor(r1 * 2699) + 1        // 1–2699 s into hour
  const faceIndex   = Math.floor(r2 * 6)               // 0–5
  // Continuous modifier rounded to nearest 1% so displayed value is clean
  const modifier    = Math.round((r3 - 0.5) * 100) / 100  // [−0.50, +0.50]

  const startMs = hourStartMs + startSecond * 1000
  const endMs   = startMs + 15 * 60 * 1000
  return {
    startMs, endMs,
    faceIndex,
    face:     DICE_FACES[faceIndex],
    modifier,
    color:    modifierColor(modifier),
  }
}

export function getDiceState(now = Date.now()) {
  const hourStart = Math.floor(now / 3_600_000) * 3_600_000
  const win    = getDiceWindowForHour(hourStart)
  const active = now >= win.startMs && now < win.endMs
  return {
    ...win,
    active,
    hourStart,
    secsLeft: active ? Math.ceil((win.endMs - now) / 1000) : 0,
  }
}

File: tools/balance/dice.py
"""Hourly dice window — bit-identical to lib/dice.ts seededRand."""

from __future__ import annotations

import math

DICE_DURATION_MS = 15 * 60 * 1000
MODIFIER_MIN = -0.50
MODIFIER_MAX = 0.50


def u32(value: int) -> int:
    return value & 0xFFFFFFFF


def js_round(value: float) -> float:
    """Match JavaScript Math.round (half toward +Infinity)."""
    return math.floor(value + 0.5)


def seeded_rand(n: int) -> float:
    s = u32(n ^ 0xDEADBEEF)
    s = u32((s ^ (s >> 16)) * 0x45D9F3B)
    s = u32((s ^ (s >> 16)) * 0x45D9F3B)
    return u32(s ^ (s >> 16)) / 0x100000000


def get_dice_window_for_hour(hour_start_ms: int) -> dict:
    seed = hour_start_ms // 3_600_000
    r1 = seeded_rand(seed * 1664525 + 1013904223)
    r3 = seeded_rand(seed * 6364136 + 1442695041)
    start_second = math.floor(r1 * 2699) + 1
    modifier = js_round((r3 - 0.5) * 100) / 100
    start_ms = hour_start_ms + start_second * 1000
    return {
        "startMs": start_ms,
        "endMs": start_ms + DICE_DURATION_MS,
        "modifier": modifier,
    }


def nftji_drop_multiplier(modifier: float) -> float:
    """README: drop rates scale by (1 + modifier)."""
    return 1.0 + modifier
